hack/docgen/schema.go

```diff
@@ -0,0 +1,242 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/iancoleman/orderedmap"
+	"github.com/invopop/jsonschema"
+	validatejsonschema "github.com/santhosh-tekuri/jsonschema/v5"
+)
+
+const ConfigSchemaURL = "https://www.talos.dev/schemas/v1alpha1/config.json"
+
+// SchemaWrapper wraps jsonschema.Schema to provide correct YAML unmarshalling using its internal JSON marshaller.
+type SchemaWrapper struct {
+	jsonschema.Schema
+}
+
+// UnmarshalYAML unmarshals the schema from YAML.
+//
+// This converts the YAML that was read from the comments to JSON,
+// then uses the custom JSON unmarshaler of the wrapped Schema.
+func (t *SchemaWrapper) UnmarshalYAML(unmarshal func(any) error) error {
+	var data map[string]any
+
+	err := unmarshal(&data)
+	if err != nil {
+		return err
+	}
+
+	jsonBytes, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+
+	return t.UnmarshalJSON(jsonBytes)
+}
+
+type SchemaTypeInfo struct {
+	typeName string
+	ref      string
+}
+
+type SchemaDefinitionInfo struct {
+	typeInfo           SchemaTypeInfo
+	arrayItemsTypeInfo *SchemaTypeInfo
+	mapValueTypeInfo   *SchemaTypeInfo
+	enumValues         []any
+}
+
+func goTypeToTypeInfo(goType string) *SchemaTypeInfo {
+	switch goType {
+	case "string":
+		return &SchemaTypeInfo{typeName: "string"}
+	case "int", "int8", "int16", "int32", "int64", "uint", "uint8", "uint16", "uint32", "uint64":
+		return &SchemaTypeInfo{typeName: "integer"}
+	case "bool":
+		return &SchemaTypeInfo{typeName: "boolean"}
+	default:
+		return &SchemaTypeInfo{ref: "#/$defs/" + goType}
+	}
+}
+
+func fieldToDefinitionInfo(field *Field) SchemaDefinitionInfo {
+	goType := field.Type
+
+	if field.Text != nil {
+		// skip enumifying boolean fields so that we don't pick up the "on", "off", "yes", "no" from the values
+		if goType != "bool" && field.Text.Values != nil {
+			enumValues := make([]any, 0, len(field.Text.Values))
+
+			for _, val := range field.Text.Values {
+				enumValues = append(enumValues, val)
+			}
+
+			return SchemaDefinitionInfo{enumValues: enumValues}
+		}
+	}
+
+	if strings.HasPrefix(goType, "[]") {
+		return SchemaDefinitionInfo{
+			typeInfo:           SchemaTypeInfo{typeName: "array"},
+			arrayItemsTypeInfo: goTypeToTypeInfo(strings.TrimPrefix(goType, "[]")),
+		}
+	}
+
+	if strings.HasPrefix(goType, "map[string]") {
+		return SchemaDefinitionInfo{
+			typeInfo:         SchemaTypeInfo{typeName: "object"},
+			mapValueTypeInfo: goTypeToTypeInfo(strings.TrimPrefix(goType, "map[string]")),
+		}
+	}
+
+	return SchemaDefinitionInfo{
+		typeInfo: *goTypeToTypeInfo(goType),
+	}
+}
+
+func typeInfoToSchema(typeInfo *SchemaTypeInfo) *jsonschema.Schema {
+	schema := &jsonschema.Schema{}
+
+	if typeInfo.typeName != "" {
+		schema.Type = typeInfo.typeName
+	}
+
+	if typeInfo.ref != "" {
+		schema.Ref = typeInfo.ref
+	}
+
+	return schema
```

Comment:
One style here and below will look better I think  - either `&schema` everywhere or `&jsonschema.Schema{}`.

Reply:
good point, done